### ISSUE TITLE
Android app contribution

### ISSUE THREAD

**Issue by iDams** [lesspass/lesspass#6]
Great tool! I am actually changing all my passwords using only lesspass.
But when I need to login to some app on my phone, it's not really easy.
Are you considering developping a mobile app that at least copy/paste the password generated?

**Comment by guillaumevincent** [comment]
Thank you for your support
The website https://lesspass.com works on mobile pretty well, and you can copy your password too.

But yeah we definitively have to improve the experience on the mobile
I'm trying right know a migration on react (because of react native) to see if the application can be ported on mobile.

**Comment by iDams** [comment]
@guillaumevincent yeah it works pretty well on mobile thanks!
Nice choice for react/react-native. If you succeed the migration, it could be very cool!

**Comment by edouard-lopez** [comment]
Desktop UX is really smooth but on mobile, I find the website workflow pretty annoying:

1. launch app ;
1. start filling form, realize I don't have password ;
1. switch/open web browser ;
1. hit address bar ;

**Comment by Rudloff** [comment]
Hello,
I started working on a simple Cordova app based on lesspass-pure: https://github.com/Rudloff/lesspass-cordova
It works exactly like the website but can be used offline.

**Comment by edouard-lopez** [comment]
Never used cordova so here is what I did:

### cordova cli

    npm install -g cordova

### Android studio (will install SDK)

    curl --location --output ~/android-studio.zip https://dl.google.com/dl/android/studio/ide-zips/2.2.2.0/android-studio-ide-145.3360264-linux.zip
    unzip android-studio.zip

Then run the installer to download SDK (~800Mo)

    ~/android-studio/bin/studio.sh
    # wait, wait, wait…

config Android SDK

    export ANDROID_HOME=~/Adnroid/Sdk/
    export PATH=$PATH:$ANDROID_HOME/tools
    export PATH=$PATH:$ANDROID_HOME/platform-tools

### Run cordova

    cordova run android

… download gradle

Download desired System Image by running: 

     ~/Android/Sdk/tools/android sdk

… work in progress so many steps to get a working env…

**Comment by edouard-lopez** [comment]
@iDams I stopped the install due to an environment too complicated to setup for the time I had (my android background is far far away…).

Could you improve your doc or provide a docker image to build the image?

**Comment by Rudloff** [comment]
@edouard-lopez Are you talking about my project?
I added some links to the Cordova docs in [the README](https://github.com/Rudloff/lesspass-cordova/blob/develop/README.md). The can also test the [prebuilt APKs](https://github.com/Rudloff/lesspass-cordova/releases).

Please open an issue if you have a problem with a specific step.

**Comment by Cheesebaron** [comment]
If you are making a mobile app, it would be much appreciated if it would work kind of like LastPass on Android, where you can have it installed as a Accessibility helper with a persistent notification, so it can fill apps too. Instead of having to switch between two apps to fill in details and copy pasting values.

**Comment by edouard-lopez** [comment]
@Rudloff thanks for the links, I will resume the install this week and test the apk.

**Comment by edouard-lopez** [comment]
Extracted @Cheesebaron suggestion to issue #50.

**Comment by charpeni** [comment]
React Native would be an excellent choice for a cross platform mobile application (iOS & Android) and provide a better feeling than Cordova.

**Comment by karlprieb** [comment]
I don't understand why people just dont use the webapp? Just create a link in your home screen and you will be good. What will change from webapp to native app?

**Comment by Cheesebaron** [comment]
@karlprieb autofill in apps could be one thing.

**Comment by Rudloff** [comment]
I created the Cordova app because I sometimes need to use it offline.
(But adding a service worker to the web app would solve this use case.)

Also I don't like that the web app loads the whole website, but that is a minor issue.

**Comment by guillaumevincent** [comment]
@karlprieb the experience with the webapp is not what we expect. We need to push forward and improve user experience. And web app will allow us to increase security because new algorithm (bcrypt argon2) are not well implemented on client side.

**Comment by karlprieb** [comment]
@Cheesebaron Yeap, it can be a problem.

@Rudloff I think we need a webpapp or app with just a form and nothing else. If we decide to use webapp we must use service workers to offline content and a webapp manifest.

@guillaumevincent It's cool migrate to bcrypt. I agree that is so slow with the JS implementation, but what I don't get is why we even consider Cordova or React Native. Both of them will run bcrypt JS implementation that is slow. Unfortunately we need a native app for a better performance :(
I still think that is good to maintain a webapp too and I can help on that :) What you think?

**Comment by charpeni** [comment]
@karlprieb Cordova run JavaScript inside a browser, but React Native will run the JavaScript with JavaScriptCore which is much faster and you can even use native library, so what's the issue ?

**Comment by karlprieb** [comment]
@charpeni Yeap, you can implement a native library bridge to React Native, I forgot that sorry.

**Comment by guillaumevincent** [comment]
@karlprieb yep native libraries build with node-gyp for example can helps us to include crypto algorithms

**Comment by sochix** [comment]
Hi, guys! I can work on react-native app. Does anyone already start thus project?

**Comment by guillaumevincent** [comment]
Hi @sochix, the initial source code is here https://github.com/lesspass/native.
Feel free to add some code. I just bootstrapped the project.

**Comment by sochix** [comment]
@guillaumevincent are you against/for using [Exponent](https://getexponent.com/) ?

**Comment by guillaumevincent** [comment]
@sochix Are the assets hosted on exponent site ? If so I do not know how far they can change the code. I do not like this idea too much. what are the advantages ?

**Comment by charpeni** [comment]
@sochix We shouldn't use Exponent since we want interaction with native librairies.

**Comment by sochix** [comment]
@guillaumevincent @charpeni as I understand, exponent just host all development environment on their servers. That's it! You can start developing in few minutes, without installing Android SDK and configuring all the mess. Everything that can be done with React Native can be done with Exponent, because it's only friendly development environment. Furthermore, if you want to switch from Exponent to local developing you can do it. Because, all code is yours and it compiles with react-native. 

Another features that I noticed:
 - easy to deploy dev versions, just share a link to an app (no pain with sending apk files)
 - you can generate apk file for submission to PlayMarket or AppStore

If I'm wrong, please correct me.

__P.s I'm not affiliated with Exponent___

**Comment by charpeni** [comment]
[How do I add custom native code to my Exponent project?](https://docs.getexponent.com/versions/v11.0.0/introduction/faq.html#how-do-i-add-custom-native-code-to-my-exponent-project)
> Right now, Exponent doesn't support custom native code, including third-party libraries which require custom native components. In an Exponent project, you never write native code--only pure JS.

**Comment by guillaumevincent** [comment]
@sochix @charpeni yes it's definitively a no go for exponent (IMO hosted components and no native code will make development harder)

**Comment by sochix** [comment]
@charpeni @guillaumevincent thanks for clarification. Will switch to bare react-native.

**Comment by sochix** [comment]
Hi guys, I have some progress :) I've made a proof-of-concept. Screenshots 
![scr1](http://vega.sochix.ru:8002/index.php/s/Bw7oJ3OfpUPSw1N/download)
![scr2](http://vega.sochix.ru:8002/index.php/s/HHzrEJHQmw6lUCK/download)

And video as well 
http://vega.sochix.ru:8002/index.php/s/wRt05Wu0z0p2GY8

**Comment by guillaumevincent** [comment]
@sochix nice work, do you want write access to https://github.com/lesspass/native/tree/master to start pushing your code ?

And we probably will need to improve consistency between app and use LessPass V2 by default.

**Comment by sochix** [comment]
> do you want write access to https://github.com/lesspass/native/tree/master to start pushing your code ?

@guillaumevincent yes, please.

Where can we have a chat about first mobile version features? (Telegram, Gitter, etc)

**Comment by guillaumevincent** [comment]
@sochix yes twitter and telegram :)

**Comment by sochix** [comment]
@guillaumevincent check [native repo](https://github.com/lesspass/native). I done development of basic features. Now app can generate V2 passwords with default pass profile.

**Comment by guillaumevincent** [comment]
@sochix I will dive in next week. Thank you for your contribution

**Comment by guillaumevincent** [comment]
@sochix I checked your code yesterday, and we need to add native pbkdf2, the 100K iterations is really too long. Crypto-js implementation seems to be the reason. I will work on this during the month. I'm moving so it can take some time. Thank you for your contribution.

**Comment by sochix** [comment]
@guillaumevincent Hi! I'll do it by myself. Can you please make a list of features that you want, except perfomance of pbkdf2?

**Comment by guillaumevincent** [comment]
@sochix I think that the mobile app should work in the notification bar https://github.com/lesspass/lesspass/issues/50

But don't spend to much time on this because the Cordova app I made is better than the react native implementation (https://play.google.com/apps/testing/com.lesspass.cordova). Based on the Cordova Hello World app, It took me 1 hour to push to Android Playstore.

The application is not fully functional, but it works and the password generation doesn't take 5 seconds.
I will spend more time on this to learn more about react native.

Btw thanks @Rudloff for your code but I took the HelloWorld from Cordova. That's why i didn't fork your repo.
